src/library_tools/library_manager.rs

```
@@ -0,0 +1,150 @@
+pub enum LibraryOptions {
+    AddNewBook,
+    BorrowBook,
+    ReturnBook,
+    ListAllBooks,
+    PrintBook,
+    Exit,
+    InvalidChoice,
+}
+
+/// Converts a u32 number to a LibraryOptions enum variant
+fn convert_u32_to_library_option(num: u32) -> LibraryOptions {
+    match num {
+        1 => LibraryOptions::AddNewBook,
+        2 => LibraryOptions::BorrowBook,
+        3 => LibraryOptions::ReturnBook,
+        4 => LibraryOptions::ListAllBooks,
+        5 => LibraryOptions::PrintBook,
+        6 => LibraryOptions::Exit,
+        _ => LibraryOptions::InvalidChoice,
+    }
+}
+
+/// LibraryManager struct that manages the library operations
+/// It contains a Library instance and methods to interact with it
+pub struct LibraryManager {
+    pub library: Library,
+}
+
+impl LibraryManager {
+    /// Constructor for the LibraryManager struct
+    /// Initializes a new LibraryManager with an empty Library
+    pub fn new() -> Self {
+        Self {
+            library: Library::new(),
+        }
+    }
+
+    /// This funciton adds a new book to the library, with data provided by the user
+    fn add_book_to_library(&mut self) {
+        println!("Enter book name:");
+        let book_name = get_string_from_user();
+        println!("Enter author name:");
+        let author_name = get_string_from_user();
+
+        self.library.add_new_book(Book::new(book_name, author_name));
+    }
+
+    /// This function borrows a book from the library
+    fn borrow_book_from_library(&mut self) {
+        println!("Enter book index to borrow:");
+        let book_index = get_u32_from_user();
+        let action_status = self.library.borrow_book(book_index);
+        match action_status {
+            Ok(_) => println!("Succesfully borrowed book!"),
+            Err(LibraryErrors::BookNotFound) => println!("No such book!"),
+            Err(LibraryErrors::InvalidBookStatus) => println!("Book is already borrowed!"),
```

Error handling should usually done in the topmost caller (i.e. the `main` function). What if I want my library to run on a far away server? There wouldn't be much sense in printing the errors, rather sending them to some client.
Bottom line is that I'd rather have these errors handled in `main` because it would be more generic, modular, and make the code cleaner.
Check out how you can propagate these errors.

---

Read about thiserror while you are at it. It is a library we use a lot in the team and is very useful for displaying and propogating errors.

---

So you think that I should not handle the errors until they reach back to the main function? I figured that main does not care as I have the library manager which I created to handle with errors because I figured that the library itself should not handle them, but this is why I created the library manager